### PR TITLE
[MIG][11.0] mozaik_phone

### PR DESCRIPTION
**Pull request by acsonefho**
- Migration of mozaik_phone
- Flake 8 (this module only)
- Pylint (this module only)
- Implements abstract unit test (from mozaik_coordinate)

### CODE REVIEW

File: odoo/addons/mozaik_coordinate/models/abstract_coordinate.py [hunk 1]
```diff
@@ -169,8 +179,8 @@ def create(self, vals):
         """
         coordinate_type = vals.get('coordinate_type')
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_coordinate/models/abstract_coordinate.py:177]: Is this line still useful ?

File: odoo/addons/mozaik_phone/data/ir_config_parameter.xml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2018 ACSONE SA/NV
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+
+<odoo>
+
+    <!-- Used by phone library -->
+    <record id="default_country_code_phone" model="ir.config_parameter">
+        <field name="key">default.country.code.phone</field>
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/data/ir_config_parameter.xml]: this key has to be moved in the other repo in the `server_environment_files/default/base.conf` file

File: odoo/addons/mozaik_phone/data/mail_message_subtype.xml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2018 ACSONE SA/NV
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+
+<odoo>
+    <record id="phone_failure_notification" model="mail.message.subtype">
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/data/mail_message_subtype.xml]: to remove as all other kind of such notification (in theory !)

File: odoo/addons/mozaik_phone/demo/phone_coordinate.xml [hunk 4]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2018 ACSONE SA/NV
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+
+<odoo>
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/demo/phone_coordinate.xml:5]: please make this file as short as possible regarding your tests

This is also a general remark for all  future modules to migrate

File: odoo/addons/mozaik_phone/demo/phone_phone.xml [hunk 5]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2018 ACSONE SA/NV
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+
+<odoo>
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/demo/phone_phone.xml:5]: id.

File: odoo/addons/mozaik_phone/models/phone_phone.py [hunk 6]
```diff
@@ -13,51 +10,68 @@
...
+    _order = 'name'
+    _unicity_keys = 'name'
 
-    @api.multi
-    def _get_linked_coordinates(self):
-        return self.env['phone.coordinate'].search(
-            [('phone_id', 'in', self.ids)])
+    name = fields.Char(
+        'Number',
+        size=50,
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/models/phone_phone.py]: to remove 

General remark: all char size can be removed

File: odoo/addons/mozaik_phone/models/phone_phone.py [hunk 7]
```diff
@@ -13,51 +10,68 @@
...
-                * if number is not parsing due to a bad encoded value
+        Number formatted into a International Number
+        If number is not starting by '+' then check if it starts by
+        '00'
+        and replace it with '+'. Otherwise set a code value with
+        a PREFIX
+        :param num: str
+        :return: str
         """
         if self.env.context.get('install_mode', False) and num[:4] == 'tbc ':
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/models/phone_phone.py]: This kind of nice-to-have should be moved into the other repo

File: odoo/addons/mozaik_phone/models/phone_phone.py [hunk 8]
```diff
@@ -67,160 +81,79 @@ def _check_and_format_number(self, num):
...
         Number, calling the _check_and_format_number method
-        :param: vals
-        :type: dictionary that contains at least 'name'
-        :return: the new phone
-        :rtype: phone.phone
+        :param vals: dict
+        :return: self recordset
         """
-        self._validate_data(vals, mode='create')
-        return super().create(vals)
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/models/phone_phone.py]: Why this form is worse ...

File: odoo/addons/mozaik_phone/models/phone_phone.py [hunk 9]
```diff
@@ -67,160 +81,79 @@ def _check_and_format_number(self, num):
...
+        self._update_values(vals, mode='create')
+        return super(PhonePhone, self).create(vals)
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/models/phone_phone.py]: ... than this one ?

File: odoo/addons/mozaik_phone/models/phone_phone.py [hunk 10]
```diff
@@ -67,160 +81,79 @@ def _check_and_format_number(self, num):
...
-        :return: True
-        :rtype: boolean
+        Update data and update the record
+        :param vals: dict
+        :return: bool
         """
-        self._validate_data(vals)
-        return super().write(vals)
+        self._update_values(vals)
+        return super(PhonePhone, self).write(vals)
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/models/phone_phone.py]: same question !

File: odoo/addons/mozaik_phone/models/res_partner.py [hunk 11]
```diff
@@ -1,134 +1,126 @@
...
+        index=True,
+    )
 
+    @api.multi
+    @api.depends(
+        'phone_coordinate_ids',
+        'phone_coordinate_ids.phone_id',
+        'phone_coordinate_ids.is_main',
+        'phone_coordinate_ids.vip',
+        'phone_coordinate_ids.unauthorized',
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/models/res_partner.py]: useful ?

File: odoo/addons/mozaik_phone/models/res_partner.py [hunk 12]
```diff
@@ -1,134 +1,126 @@
...
+        'phone_coordinate_ids.active',
+        'phone_coordinate_inactive_ids.active',
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/models/res_partner.py]: useful ? Maybe a former bug... Still relevant?

File: odoo/addons/mozaik_phone/models/res_partner.py [hunk 13]
```diff
@@ -1,134 +1,126 @@
...
+            ('active', '=', True),
+        ]
+        all_phone_coordinates = self.env['phone.coordinate'].sudo().search(
+            domain)
+        vip = 'VIP'
+        for record in self:
+            # Get coordinate phone depending of the type(fix, fax, mobile)
+            phone_coordinates = all_phone_coordinates.filtered(
+                lambda c, p=record: c.partner_id.id == p.id)
+            phone_coordinates = phone_coordinates.with_prefetch(self._prefetch)
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/models/res_partner.py]: `self._prefetch` or `record._prefetch` or `phone_coordinates._prefetch` ?

File: odoo/addons/mozaik_phone/tests/test_change_phone_type.py [hunk 14]
```diff
@@ -1,141 +1,110 @@
...
+        """
+        Trying to change mobile5 (not main) from mobile to fix and
+        set is as main in the new category.
+        Expected results are:
+        - mobile5 type should be FIX
+        - mobile_coordinate_for_jacques_2 type should be FIX and MAIN
+        - mobile_coordinate_for_jacques_1 should remain main
+        - fix_coordinate_for_jacques_1 should not be main anymore
+        """
+        wizard = self.wizard_obj.create({
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/tests/test_change_phone_type.py]: Is a `create()` mandatory here ? A `new()` is maybe enough !

General remark for wizards

File: odoo/addons/mozaik_phone/tests/test_res_partner.py [hunk 15]
```diff
@@ -1,233 +1,64 @@
...
-            uid,
-            self.phone_coordinate_id_1,
-            context=context).phone_id.name
-        self.assertEqual(
-            partner_value['phone'] == phone_value,
-            True,
+        self.coordinate1.write({
+            'phone_id': self.phone2.id,
+        })
+        self.assertEquals(
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/tests/test_res_partner.py]: deprecated

File: odoo/addons/mozaik_phone/views/phone_coordinate.xml [hunk 16]
```diff
@@ -4,43 +4,35 @@
 
 <odoo>
 
-    <record model="ir.ui.menu" id="phone_coordinate_submenu">
-        <field name="name">Phone Coordinates</field>
-        <field name="parent_id" ref="contacts.menu_contacts"/> <!--TODO mozaik_base.persons_root_menu-->
-        <field name="sequence" eval="40"/>
-    </record>
+    <menuitem id="phone_coordinate_submenu" name="Phone Coordinates"
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/views/phone_coordinate.xml]: Why this rollback to menuitem ?

General remark: no use of `<menuitem/>` if possible

File: odoo/addons/mozaik_phone/views/phone_coordinate.xml [hunk 17]
```diff
@@ -4,43 +4,35 @@
...
-            <xpath expr="//group[@name='moreinfo']" position="inside">
-                <span name="anchor_for_inherit_view" />
-            </xpath>
+                <attribute name="context" translation="off">{
+                    'search_default_duplicate': True,
+                    'search_default_phone_id': phone_id,
+                    'default_phone_id': phone_id,
+                    }</attribute>
+            </button>
+            <group name="moreinfo" position="inside">
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/views/phone_coordinate.xml:36]: look in `mozaik_structure` how I suggest to keep this form section... as a `<page/>` of a `<notebook/>`

File: odoo/addons/mozaik_phone/views/phone_coordinate.xml [hunk 18]
```diff
@@ -71,15 +61,15 @@
         <field name="model">phone.coordinate</field>
         <field name="inherit_id" ref="mozaik_coordinate.abstract_coordinate_tree_view" />
         <field name="arch" type="xml">
-            <xpath expr="//field[@name='partner_id']" position="after">
+            <field name="partner_id" position="after">
                 <field name="phone_id" />
-                <field name="coordinate_type" /> <!--TODO groups="mozaik_base.res_groups_developper" />-->
-            </xpath>
+                <field name="coordinate_type" groups="base.group_erp_manager"/>
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/views/phone_coordinate.xml]: IMO, not useful in the tree.

File: odoo/addons/mozaik_phone/views/phone_phone.xml [hunk 19]
```diff
@@ -59,14 +58,13 @@
...
-                                    <field name="vip" /><!--TODO groups="mozaik_base.mozaik_res_groups_vip_reader" />-->
+                                    <field name="vip" groups="mozaik_coordinate.res_groups_coordinate_vip_reader"/>
                                 </tree>
                             </field>
                         </page>
                     </notebook>
                 </sheet>
-                <!-- TODO -->
-                <div name="dev" class="oe_form_sheetbg"> <!--TODO groups="mozaik_base.res_groups_developper">-->
+                <div name="dev" class="oe_form_sheetbg" groups="base.group_erp_manager">
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/views/phone_phone.xml]: Put it in the notebook

File: odoo/addons/mozaik_phone/views/res_partner.xml [hunk 20]
```diff
@@ -4,51 +4,54 @@
 
 <odoo>
 
-    <record model="ir.ui.view" id="res_partner_form_view">
+    <record model="ir.ui.view" id="res_partner_form_2_view">
         <field name="name">res.partner.form (in mozaik_phone)</field>
         <field name="model">res.partner</field>
-        <field name="inherit_id" ref="base.view_partner_form"/>
+        <field name="inherit_id" ref="mozaik_coordinate.res_partner_form_view"/>
+        <field name="priority" eval="150"/>
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/views/res_partner.xml]: Useful ?

File: odoo/addons/mozaik_phone/wizards/change_main_phone.py [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+
+class ChangeMainPhone(models.TransientModel):
+    _name = 'change.main.phone'
+    _inherit = 'change.main.coordinate'
+    _description = 'Change Main Phone Wizard'
+
+    old_phone_id = fields.Many2one(
+        "phone.phone",
+        "Current main phone",
+        ondelete="cascade",
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/wizards/change_main_phone.py]: not necessary for a non required M2O

File: odoo/addons/mozaik_phone/wizards/change_main_phone.py [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+    @api.onchange('partner_id', 'phone_id')
+    def _onchange_phone_id(self):
+        if self.phone_id.type == 'fix':
+            self.old_phone_id = self.partner_id.fix_coordinate_id.phone_id.id
+        elif self.phone_id.type == 'fax':
+            self.old_phone_id = self.partner_id.fax_coordinate_id.phone_id.id
+        else:
+            self.old_phone_id = self.partner_id.mobile_coordinate_id\
+                .phone_id.id
+        self.change_allowed = not bool(self.phone_id == self.old_phone_id)
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/wizards/change_main_phone.py]: bool is unless

File: odoo/addons/mozaik_phone/wizards/change_main_phone.py [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+        Based on the context, fill the partner_id automatically.
+        If the mode is switch, auto-fill the phone_id
+        :param fields_list: list of str
+        :return: dict
+        """
+        result = super(ChangeMainPhone, self).default_get(fields_list)
+        context = self.env.context
+        active_id = context.get('active_ids', context.get('active_id'))
+        if active_id and isinstance(active_id, list):
+            active_id = active_id[0]
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/wizards/change_main_phone.py:38]: be careful `active_id` may be a phone coordinate

File: odoo/addons/mozaik_phone/wizards/change_main_phone.xml [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+        <field name="name">change.main.phone.wizard</field>
+        <field name="model">change.main.phone</field>
+        <field name="inherit_id" ref="mozaik_coordinate.change_main_coordinate_form" />
+        <field name="arch" type="xml">
+
+            <xpath expr="//field[@name='invalidate_previous_coordinate']" position="before">
+                <field name="old_phone_id" options='{"no_open":True}' readonly='1'/>
+                <field name="phone_id" options='{"no_open":True}'
+                       readonly="context.get('mode', '') == 'switch'"
+                       on_change="onchange_phone_id(phone_id, partner_id)"/>
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/wizards/change_main_phone.xml]: to remove

File: odoo/addons/mozaik_phone/wizards/change_phone_type.xml [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+<!-- Copyright 2018 ACSONE SA/NV
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+
+<odoo>
+
+    <record id="change_phone_type_form" model="ir.ui.view">
+        <field name="name">change.phone.type.form</field>
+        <field name="model">change.phone.type</field>
+        <field name="arch" type="xml">
+            <form string="Change Phone Type">
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/wizards/change_phone_type.xml]: string can be removed

File: odoo/addons/mozaik_phone/wizards/change_phone_type.xml [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+                    or
+                    <button string="Cancel" class="oe_link" special="cancel" />
+                </footer>
+            </form>
+        </field>
+    </record>
+
+    <record id="change_phone_type_action" model="ir.actions.act_window">
+        <field name="name">Change Phone Type</field>
+        <field name="view_type">form</field>
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/wizards/change_phone_type.xml]: to remove

General remark: view_type can always be removed

File: odoo/addons/mozaik_phone/wizards/failure_editor_view.xml [hunk 27]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2018 ACSONE SA/NV
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+
+<odoo>
+
+    <record model="ir.actions.act_window" id="failure_editor_act_window">
+        <field name="name">Failure Editor</field>
+        <field name="view_type">form</field>
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/wizards/failure_editor_view.xml]: id.

File: odoo/addons/mozaik_phone/wizards/failure_editor_view.xml [hunk 28]
```diff
@@ -0,0 +1,20 @@
...
+        <field name="view_mode">form</field>
+        <field name="res_model">failure.editor</field>
+        <field name="multi" eval="0"/>
```
Comment (Olivier-LAURENT) [inline odoo/addons/mozaik_phone/wizards/failure_editor_view.xml]: useful ?

**Review by Olivier-LAURENT — CHANGES_REQUESTED** [review]

File: odoo/addons/mozaik_coordinate/models/abstract_coordinate.py (hunk 1 above)
Comment (RobinetDenisAcsone) [reply odoo/addons/mozaik_coordinate/models/abstract_coordinate.py:177]: coordinate_type is used later inn the create, if the coordinate is main

File: odoo/addons/mozaik_phone/models/res_partner.py (hunk 12 above)
Comment (RobinetDenisAcsone) [reply odoo/addons/mozaik_phone/models/res_partner.py]: It doesn't seems to be useful, I remove it

**Pull request merged**